Integrate vector maths library

in `Clang`, there is `-fveclib=` flag that allows to specify the vector library. Since `exp` and other standard LLVM intrinsics are not fully vectorisable, this ticket is to:

- Add an option to take a specific (we are interested in SVML) vector library for LLVM visitor
- If SVML flag is set, instead of LLVM intrinsics generate calls to functions from SVML.

Additionally, modify JIT to link against shared libraries.

**Note** The last option may not be needed! It might be the case that LLVM has API to change `llvm.intr.exp` calls into `__svml_exp128` or other vector intrinsics.

Inspiration: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/TargetLibraryInfo.cpp#L20

> update LLVM to trunk?

I will take care of that in our CI.

I quickly build against trunk/v13. NMODL builds fine with some warnings and all tests are running fine:

```bash
$ cmake .. -DCMAKE_INSTALL_PREFIX=`pwd`/install -DCMAKE_CXX_FLAGS="-g -O3" -DNMODL_FORMATTING=ON -DLLVM_DIR=/usr/lib/llvm-13/share/llvm/cmake/
...
/home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp: In member function ‘llvm::Value* nmodl::codegen::CodegenLLVMVisitor::codegen_instance_var(const nmodl::ast::CodegenInstanceVar&)’:
/home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp:83:73: warning: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Value*, const llvm::Twine&)’ is deprecated: Use the version that explicitly specifies the loaded type instead [-Wdeprecated-declarations]
   83 |     llvm::Value* instance_ptr = builder.CreateLoad(lookup(instance_name));
      |                                                                         ^
In file included from /usr/lib/llvm-13/include/llvm/Support/ErrorHandling.h:17,
                 from /usr/lib/llvm-13/include/llvm/ADT/Hashing.h:48,
                 from /usr/lib/llvm-13/include/llvm/ADT/ArrayRef.h:12,
                 from /usr/lib/llvm-13/include/llvm/IR/IRBuilder.h:18,
                 from /home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.hpp:26,
                 from /home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp:8:
/usr/lib/llvm-13/include/llvm/IR/IRBuilder.h:1660:39: note: declared here
 1660 |   LLVM_ATTRIBUTE_DEPRECATED(LoadInst *CreateLoad(Value *Ptr,
      |                                       ^~~~~~~~~~
/usr/lib/llvm-13/include/llvm/Support/Compiler.h:320:74: note: in definition of macro ‘LLVM_ATTRIBUTE_DEPRECATED’
  320 | #define LLVM_ATTRIBUTE_DEPRECATED(decl, message) [[deprecated(message)]] decl
      |                                                                          ^~~~
/home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp: In member function ‘llvm::Value* nmodl::codegen::CodegenLLVMVisitor::get_array_index(const nmodl::ast::IndexedName&)’:
/home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp:149:45: warning: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Value*, const llvm::Twine&)’ is deprecated: Use the version that explicitly specifies the loaded type instead [-Wdeprecated-declarations]
  149 |         index_value = builder.CreateLoad(ptr);
      |                                             ^
In file included from /usr/lib/llvm-13/include/llvm/Support/ErrorHandling.h:17,
                 from /usr/lib/llvm-13/include/llvm/ADT/Hashing.h:48,
                 from /usr/lib/llvm-13/include/llvm/ADT/ArrayRef.h:12,
                 from /usr/lib/llvm-13/include/llvm/IR/IRBuilder.h:18,
                 from /home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.hpp:26,
                 from /home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp:8:
/usr/lib/llvm-13/include/llvm/IR/IRBuilder.h:1660:39: note: declared here
 1660 |   LLVM_ATTRIBUTE_DEPRECATED(LoadInst *CreateLoad(Value *Ptr,
      |                                       ^~~~~~~~~~
/usr/lib/llvm-13/include/llvm/Support/Compiler.h:320:74: note: in definition of macro ‘LLVM_ATTRIBUTE_DEPRECATED’
  320 | #define LLVM_ATTRIBUTE_DEPRECATED(decl, message) [[deprecated(message)]] decl
      |                                                                          ^~~~
/home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp: In member function ‘virtual void nmodl::codegen::CodegenLLVMVisitor::visit_codegen_return_statement(const nmodl::ast::CodegenReturnStatement&)’:
/home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp:714:60: warning: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Value*, const llvm::Twine&)’ is deprecated: Use the version that explicitly specifies the loaded type instead [-Wdeprecated-declarations]
  714 |     llvm::Value* ret_value = builder.CreateLoad(lookup(ret));
      |                                                            ^
In file included from /usr/lib/llvm-13/include/llvm/Support/ErrorHandling.h:17,
                 from /usr/lib/llvm-13/include/llvm/ADT/Hashing.h:48,
                 from /usr/lib/llvm-13/include/llvm/ADT/ArrayRef.h:12,
                 from /usr/lib/llvm-13/include/llvm/IR/IRBuilder.h:18,
                 from /home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.hpp:26,
                 from /home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp:8:
/usr/lib/llvm-13/include/llvm/IR/IRBuilder.h:1660:39: note: declared here
 1660 |   LLVM_ATTRIBUTE_DEPRECATED(LoadInst *CreateLoad(Value *Ptr,
      |                                       ^~~~~~~~~~
/usr/lib/llvm-13/include/llvm/Support/Compiler.h:320:74: note: in definition of macro ‘LLVM_ATTRIBUTE_DEPRECATED’
  320 | #define LLVM_ATTRIBUTE_DEPRECATED(decl, message) [[deprecated(message)]] decl
      |                                                                          ^~~~
/home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp: In member function ‘virtual void nmodl::codegen::CodegenLLVMVisitor::visit_var_name(const nmodl::ast::VarName&)’:
/home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp:938:78: warning: ‘llvm::LoadInst* llvm::IRBuilderBase::CreateLoad(llvm::Value*, const llvm::Twine&)’ is deprecated: Use the version that explicitly specifies the loaded type instead [-Wdeprecated-declarations]
  938 |     llvm::Value* var = ptr->getType()->isPointerTy() ? builder.CreateLoad(ptr) : ptr;
      |                                                                              ^
In file included from /usr/lib/llvm-13/include/llvm/Support/ErrorHandling.h:17,
                 from /usr/lib/llvm-13/include/llvm/ADT/Hashing.h:48,
                 from /usr/lib/llvm-13/include/llvm/ADT/ArrayRef.h:12,
                 from /usr/lib/llvm-13/include/llvm/IR/IRBuilder.h:18,
                 from /home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.hpp:26,
                 from /home/kumbhar/workarena/repos/bbp/nmodl/src/codegen/llvm/codegen_llvm_visitor.cpp:8:
/usr/lib/llvm-13/include/llvm/IR/IRBuilder.h:1660:39: note: declared here
 1660 |   LLVM_ATTRIBUTE_DEPRECATED(LoadInst *CreateLoad(Value *Ptr,
      |                                       ^~~~~~~~~~
/usr/lib/llvm-13/include/llvm/Support/Compiler.h:320:74: note: in definition of macro ‘LLVM_ATTRIBUTE_DEPRECATED’
  320 | #define LLVM_ATTRIBUTE_DEPRECATED(decl, message) [[deprecated(message)]] decl
...
$ make test
...
100% tests passed, 0 tests failed out of 95
```

> Create a placeholder that just replaces the names in our generation of exp, pow, etc?

Do you mean placeholder that will run `VecLibReplace` pass with appropriate library (svml) and that will do work for us?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -149,6 +149,9 @@ if(NMODL_ENABLE_LLVM)
   include(LLVMHelper)
   include_directories(${LLVM_INCLUDE_DIRS})
   add_definitions(-DNMODL_LLVM_BACKEND)
+  if(LLVM_VERSION VERSION_LESS_EQUAL 12)
+    add_definitions(-DLLVM_VERSION_LESS_THAN_13)
+  endif()
 endif()
 
 # =============================================================================

--- a/cmake/LLVMHelper.cmake
+++ b/cmake/LLVMHelper.cmake
@@ -7,8 +7,11 @@ find_package(LLVM REQUIRED CONFIG)
 # include LLVM header and core library
 llvm_map_components_to_libnames(
   LLVM_LIBS_TO_LINK
+  analysis
+  codegen
   core
   instcombine
+  mc
   native
   orcjit
   scalaropts

--- a/src/codegen/llvm/codegen_llvm_visitor.cpp
+++ b/src/codegen/llvm/codegen_llvm_visitor.cpp
@@ -11,6 +11,8 @@
 #include "visitors/rename_visitor.hpp"
 #include "visitors/visitor_utils.hpp"
 
+#include "llvm/Analysis/TargetLibraryInfo.h"
+#include "llvm/CodeGen/ReplaceWithVeclib.h"
 #include "llvm/IR/AssemblyAnnotationWriter.h"
 #include "llvm/IR/BasicBlock.h"
 #include "llvm/IR/Constants.h"
@@ -19,6 +21,7 @@
 #include "llvm/IR/Type.h"
 #include "llvm/IR/ValueSymbolTable.h"
 #include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Host.h"
 #include "llvm/Support/ToolOutputFile.h"
 
 namespace nmodl {
@@ -292,21 +295,21 @@ std::shared_ptr<ast::InstanceStruct> CodegenLLVMVisitor::get_instance_struct_ptr
     return instance_var_helper.instance;
 }
 
-void CodegenLLVMVisitor::run_llvm_opt_passes() {
+void CodegenLLVMVisitor::run_ir_opt_passes() {
     /// run some common optimisation passes that are commonly suggested
-    fpm.add(llvm::createInstructionCombiningPass());
-    fpm.add(llvm::createReassociatePass());
-    fpm.add(llvm::createGVNPass());
-    fpm.add(llvm::createCFGSimplificationPass());
+    opt_pm.add(llvm::createInstructionCombiningPass());
+    opt_pm.add(llvm::createReassociatePass());
+    opt_pm.add(llvm::createGVNPass());
+    opt_pm.add(llvm::createCFGSimplificationPass());
 
     /// initialize pass manager
-    fpm.doInitialization();
+    opt_pm.doInitialization();
 
     /// iterate over all functions and run the optimisation passes
     auto& functions = module->getFunctionList();
     for (auto& function: functions) {
         llvm::verifyFunction(function);
-        fpm.run(function);
+        opt_pm.run(function);
     }
 }
 
@@ -892,7 +895,28 @@ void CodegenLLVMVisitor::visit_program(const ast::Program& node) {
 
     if (opt_passes) {
         logger->info("Running LLVM optimisation passes");
-        run_llvm_opt_passes();
+        run_ir_opt_passes();
+    }
+
+    // Optionally, replace LLVM's maths intrinsics with vector library calls.
+    if (vector_library != llvm::TargetLibraryInfoImpl::NoLibrary) {
+        // First, get the target library information.
+        llvm::Triple triple(llvm::sys::getDefaultTargetTriple());
+        llvm::TargetLibraryInfoImpl target_lib_info = llvm::TargetLibraryInfoImpl(triple);
+
+        if (vector_library != llvm::TargetLibraryInfoImpl::LIBMVEC_X86 ||
+            (triple.isX86() && triple.isArch64Bit())) {
+            target_lib_info.addVectorizableFunctionsFromVecLib(vector_library);
+        }
+
+        // Run the codegen optimisation passes.
+        codegen_pm.add(new llvm::TargetLibraryInfoWrapperPass(target_lib_info));
+        codegen_pm.add(new llvm::ReplaceWithVeclibLegacy);
+        codegen_pm.doInitialization();
+        for (auto& function: module->getFunctionList()) {
+            if (!function.isDeclaration())
+                codegen_pm.run(function);
+        }
     }
 
     // If the output directory is specified, save the IR to .ll file.

--- a/src/codegen/llvm/codegen_llvm_visitor.cpp
+++ b/src/codegen/llvm/codegen_llvm_visitor.cpp
@@ -12,7 +12,6 @@
 #include "visitors/visitor_utils.hpp"
 
 #include "llvm/Analysis/TargetLibraryInfo.h"
-#include "llvm/CodeGen/ReplaceWithVeclib.h"
 #include "llvm/IR/AssemblyAnnotationWriter.h"
 #include "llvm/IR/BasicBlock.h"
 #include "llvm/IR/Constants.h"
@@ -23,6 +22,10 @@
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/Host.h"
 #include "llvm/Support/ToolOutputFile.h"
+
+#ifndef LLVM_VERSION_LESS_THAN_13
+#include "llvm/CodeGen/ReplaceWithVeclib.h"
+#endif
 
 namespace nmodl {
 namespace codegen {
@@ -899,7 +902,12 @@ void CodegenLLVMVisitor::visit_program(const ast::Program& node) {
     }
 
     // Optionally, replace LLVM's maths intrinsics with vector library calls.
-    if (vector_library != llvm::TargetLibraryInfoImpl::NoLibrary) {
+    if (vector_width > 1 && vector_library != llvm::TargetLibraryInfoImpl::NoLibrary) {
+#ifdef LLVM_VERSION_LESS_THAN_13
+        logger->warn(
+            "This version of LLVM does not support replacement of LLVM intrinsics with vector "
+            "library calls");
+#else
         // First, get the target library information.
         llvm::Triple triple(llvm::sys::getDefaultTargetTriple());
         llvm::TargetLibraryInfoImpl target_lib_info = llvm::TargetLibraryInfoImpl(triple);
@@ -920,6 +928,7 @@ void CodegenLLVMVisitor::visit_program(const ast::Program& node) {
                 codegen_pm.run(function);
         }
         codegen_pm.doFinalization();
+#endif
     }
 
     // If the output directory is specified, save the IR to .ll file.

--- a/src/codegen/llvm/codegen_llvm_visitor.cpp
+++ b/src/codegen/llvm/codegen_llvm_visitor.cpp
@@ -904,12 +904,14 @@ void CodegenLLVMVisitor::visit_program(const ast::Program& node) {
         llvm::Triple triple(llvm::sys::getDefaultTargetTriple());
         llvm::TargetLibraryInfoImpl target_lib_info = llvm::TargetLibraryInfoImpl(triple);
 
+        // Populate target library information with vectorisable functions. Since libmvec is
+        // supported for x86_64 only, have a check to catch other architectures.
         if (vector_library != llvm::TargetLibraryInfoImpl::LIBMVEC_X86 ||
             (triple.isX86() && triple.isArch64Bit())) {
             target_lib_info.addVectorizableFunctionsFromVecLib(vector_library);
         }
 
-        // Run the codegen optimisation passes.
+        // Run the codegen optimisation passes that replace maths intrinsics.
         codegen_pm.add(new llvm::TargetLibraryInfoWrapperPass(target_lib_info));
         codegen_pm.add(new llvm::ReplaceWithVeclibLegacy);
         codegen_pm.doInitialization();

--- a/src/codegen/llvm/codegen_llvm_visitor.cpp
+++ b/src/codegen/llvm/codegen_llvm_visitor.cpp
@@ -917,6 +917,7 @@ void CodegenLLVMVisitor::visit_program(const ast::Program& node) {
             if (!function.isDeclaration())
                 codegen_pm.run(function);
         }
+        codegen_pm.doFinalization();
     }
 
     // If the output directory is specified, save the IR to .ll file.

--- a/src/codegen/llvm/codegen_llvm_visitor.hpp
+++ b/src/codegen/llvm/codegen_llvm_visitor.hpp
@@ -49,7 +49,9 @@ namespace codegen {
 /// A map to query vector library by its string value.
 static const std::map<std::string, llvm::TargetLibraryInfoImpl::VectorLibrary> veclib_map = {
             {"Accelerate", llvm::TargetLibraryInfoImpl::Accelerate},
+#ifndef LLVM_VERSION_LESS_THAN_13
             {"libmvec", llvm::TargetLibraryInfoImpl::LIBMVEC_X86},
+#endif
             {"MASSV", llvm::TargetLibraryInfoImpl::MASSV},
             {"SVML", llvm::TargetLibraryInfoImpl::SVML},
             {"none", llvm::TargetLibraryInfoImpl::NoLibrary}

--- a/src/codegen/llvm/codegen_llvm_visitor.hpp
+++ b/src/codegen/llvm/codegen_llvm_visitor.hpp
@@ -23,6 +23,7 @@
 #include "utils/logger.hpp"
 #include "visitors/ast_visitor.hpp"
 
+#include "llvm/Analysis/TargetLibraryInfo.h"
 #include "llvm/IR/IRBuilder.h"
 #include "llvm/IR/LLVMContext.h"
 #include "llvm/IR/LegacyPassManager.h"
@@ -45,6 +46,15 @@ namespace codegen {
  * @{
  */
 
+/// A map to query vector library by its string value.
+static const std::map<std::string, llvm::TargetLibraryInfoImpl::VectorLibrary> veclib_map = {
+            {"Accelerate", llvm::TargetLibraryInfoImpl::Accelerate},
+            {"libmvec", llvm::TargetLibraryInfoImpl::LIBMVEC_X86},
+            {"MASSV", llvm::TargetLibraryInfoImpl::MASSV},
+            {"SVML", llvm::TargetLibraryInfoImpl::SVML},
+            {"none", llvm::TargetLibraryInfoImpl::NoLibrary}
+};
+
 /**
  * \class CodegenLLVMVisitor
  * \brief %Visitor for transforming NMODL AST to LLVM IR
@@ -65,7 +75,14 @@ class CodegenLLVMVisitor: public visitor::ConstAstVisitor {
 
     llvm::IRBuilder<> builder;
 
-    llvm::legacy::FunctionPassManager fpm;
+    // Pass manager for optimisation passes that are used for target code generation.
+    llvm::legacy::FunctionPassManager codegen_pm;
+
+    // Vector library used for maths functions.
+    llvm::TargetLibraryInfoImpl::VectorLibrary vector_library;
+
+    // Pass manager for optimisation passes that are run on IR and are not related to target.
+    llvm::legacy::FunctionPassManager opt_pm;
 
     // Stack to hold visited values
     std::vector<llvm::Value*> values;
@@ -97,7 +114,7 @@ class CodegenLLVMVisitor: public visitor::ConstAstVisitor {
      * LLVM provides number of optimisation passes that can be run on the generated IR.
      * Here we run common optimisation LLVM passes that benefits code optimisation.
      */
-    void run_llvm_opt_passes();
+    void run_ir_opt_passes();
 
   public:
     /**
@@ -110,14 +127,17 @@ class CodegenLLVMVisitor: public visitor::ConstAstVisitor {
                        const std::string& output_dir,
                        bool opt_passes,
                        bool use_single_precision = false,
-                       int vector_width = 1)
+                       int vector_width = 1,
+                       std::string vec_lib = "none")
         : mod_filename(mod_filename)
         , output_dir(output_dir)
         , opt_passes(opt_passes)
         , use_single_precision(use_single_precision)
         , vector_width(vector_width)
+        , vector_library(veclib_map.at(vec_lib))
         , builder(*context)
-        , fpm(module.get()) {}
+        , codegen_pm(module.get())
+        , opt_pm(module.get()) {}
 
 
     /**

--- a/src/codegen/llvm/codegen_llvm_visitor.hpp
+++ b/src/codegen/llvm/codegen_llvm_visitor.hpp
@@ -48,14 +48,13 @@ namespace codegen {
 
 /// A map to query vector library by its string value.
 static const std::map<std::string, llvm::TargetLibraryInfoImpl::VectorLibrary> veclib_map = {
-            {"Accelerate", llvm::TargetLibraryInfoImpl::Accelerate},
+    {"Accelerate", llvm::TargetLibraryInfoImpl::Accelerate},
 #ifndef LLVM_VERSION_LESS_THAN_13
-            {"libmvec", llvm::TargetLibraryInfoImpl::LIBMVEC_X86},
+    {"libmvec", llvm::TargetLibraryInfoImpl::LIBMVEC_X86},
 #endif
-            {"MASSV", llvm::TargetLibraryInfoImpl::MASSV},
-            {"SVML", llvm::TargetLibraryInfoImpl::SVML},
-            {"none", llvm::TargetLibraryInfoImpl::NoLibrary}
-};
+    {"MASSV", llvm::TargetLibraryInfoImpl::MASSV},
+    {"SVML", llvm::TargetLibraryInfoImpl::SVML},
+    {"none", llvm::TargetLibraryInfoImpl::NoLibrary}};
 
 /**
  * \class CodegenLLVMVisitor

--- a/src/nmodl/main.cpp
+++ b/src/nmodl/main.cpp
@@ -175,10 +175,13 @@ int main(int argc, const char* argv[]) {
     bool llvm_float_type(false);
 
     /// run llvm optimisation passes
-    bool llvm_opt_passes(false);
+    bool llvm_ir_opt_passes(false);
 
     /// llvm vector width
     int llvm_vec_width = 1;
+
+    /// vector library
+    std::string vec_lib("none");
 
     /// run llvm benchmark
     bool run_benchmark(false);
@@ -308,14 +311,17 @@ int main(int argc, const char* argv[]) {
         llvm_ir,
         "Generate LLVM IR ({})"_format(llvm_ir))->ignore_case();
     llvm_opt->add_flag("--opt",
-        llvm_opt_passes,
-        "Run LLVM optimisation passes ({})"_format(llvm_opt_passes))->ignore_case();
+                       llvm_ir_opt_passes,
+                       "Run LLVM optimisation passes ({})"_format(llvm_ir_opt_passes))->ignore_case();
     llvm_opt->add_flag("--single-precision",
                        llvm_float_type,
                        "Use single precision floating-point types ({})"_format(llvm_float_type))->ignore_case();
     llvm_opt->add_option("--vector-width",
         llvm_vec_width,
         "LLVM explicit vectorisation width ({})"_format(llvm_vec_width))->ignore_case();
+    llvm_opt->add_option("--veclib",
+                         vec_lib,
+                         "Vector library for maths functions ({})"_format(vec_lib))->check(CLI::IsMember({"Accelerate", "libmvec", "MASSV", "SVML", "none"}));
 
     // LLVM IR benchmark options.
     auto benchmark_opt = app.add_subcommand("benchmark", "LLVM benchmark option")->ignore_case();
@@ -330,7 +336,7 @@ int main(int argc, const char* argv[]) {
                        "Number of experiments for benchmarking ({})"_format(repeat))->ignore_case();
     benchmark_opt->add_option("--backend",
                        backend,
-                       "Target's backend ({})"_format(backend))->ignore_case()->check(CLI::IsMember({"avx2", "default", "sse2"}));;
+                       "Target's backend ({})"_format(backend))->ignore_case()->check(CLI::IsMember({"avx2", "default", "sse2"}));
 #endif
     // clang-format on
 
@@ -620,7 +626,7 @@ int main(int argc, const char* argv[]) {
 
             if (run_benchmark) {
                 logger->info("Running LLVM benchmark");
-                benchmark::LLVMBuildInfo info{llvm_vec_width, llvm_opt_passes, llvm_float_type};
+                benchmark::LLVMBuildInfo info{llvm_vec_width, llvm_ir_opt_passes, llvm_float_type};
                 benchmark::LLVMBenchmark bench(
                     modfile, output_dir, info, repeat, instance_size, backend);
                 bench.benchmark(ast);
@@ -628,8 +634,12 @@ int main(int argc, const char* argv[]) {
 
             else if (llvm_ir) {
                 logger->info("Running LLVM backend code generator");
-                CodegenLLVMVisitor visitor(
-                    modfile, output_dir, llvm_opt_passes, llvm_float_type, llvm_vec_width);
+                CodegenLLVMVisitor visitor(modfile,
+                                           output_dir,
+                                           llvm_ir_opt_passes,
+                                           llvm_float_type,
+                                           llvm_vec_width,
+                                           vec_lib);
                 visitor.visit_program(*ast);
                 ast_to_nmodl(*ast, filepath("llvm", "mod"));
                 ast_to_json(*ast, filepath("llvm", "json"));

--- a/test/unit/codegen/codegen_llvm_ir.cpp
+++ b/test/unit/codegen/codegen_llvm_ir.cpp
@@ -1097,6 +1097,7 @@ SCENARIO("Vector library calls", "[visitor][llvm][vector_lib]") {
             REQUIRE(std::regex_search(no_library_module_str, m, exp_decl));
             REQUIRE(std::regex_search(no_library_module_str, m, exp_call));
 
+#ifndef LLVM_VERSION_LESS_THAN_13
             // Check exponential calls are replaced with calls to SVML library.
             std::string svml_library_module_str = run_llvm_visitor(nmodl_text,
                                                                    /*opt=*/false,
@@ -1144,6 +1145,7 @@ SCENARIO("Vector library calls", "[visitor][llvm][vector_lib]") {
             REQUIRE(std::regex_search(accelerate_library_module_str, m, accelerate_exp_decl));
             REQUIRE(std::regex_search(accelerate_library_module_str, m, accelerate_exp_call));
             REQUIRE(!std::regex_search(accelerate_library_module_str, m, fexp_call));
+#endif
         }
     }
 }

--- a/test/unit/codegen/codegen_llvm_ir.cpp
+++ b/test/unit/codegen/codegen_llvm_ir.cpp
@@ -36,7 +36,8 @@ using nmodl::parser::NmodlDriver;
 std::string run_llvm_visitor(const std::string& text,
                              bool opt = false,
                              bool use_single_precision = false,
-                             int vector_width = 1) {
+                             int vector_width = 1,
+                             std::string vec_lib = "none") {
     NmodlDriver driver;
     const auto& ast = driver.parse_string(text);
 
@@ -48,7 +49,8 @@ std::string run_llvm_visitor(const std::string& text,
                                              /*output_dir=*/".",
                                              opt,
                                              use_single_precision,
-                                             vector_width);
+                                             vector_width,
+                                             vec_lib);
     llvm_visitor.visit_program(*ast);
     return llvm_visitor.dump_module();
 }
@@ -1052,6 +1054,99 @@ SCENARIO("Vectorised derivative block", "[visitor][llvm][derivative]") {
 
             auto epilogue_loop = reindent_text(to_nmodl(result[1]));
             REQUIRE(epilogue_loop == reindent_text(expected_epilogue_loop));
+        }
+    }
+}
+
+//=============================================================================
+// Vector library calls.
+//=============================================================================
+
+SCENARIO("Vector library calls", "[visitor][llvm][vector_lib]") {
+    GIVEN("A vector LLVM intrinsic") {
+        std::string nmodl_text = R"(
+            NEURON {
+                SUFFIX hh
+                NONSPECIFIC_CURRENT il
+                RANGE minf, mtau
+            }
+            STATE {
+                m
+            }
+            ASSIGNED {
+                v (mV)
+                minf
+                mtau (ms)
+            }
+            BREAKPOINT {
+                SOLVE states METHOD cnexp
+                il = 2
+            }
+            DERIVATIVE states {
+                m' = (minf-m) / mtau
+            }
+        )";
+
+        THEN("it is replaced with an appropriate vector library call") {
+            std::smatch m;
+
+            // Check exponential intrinsic is created.
+            std::string no_library_module_str = run_llvm_visitor(nmodl_text,
+                                                                 /*opt=*/false,
+                                                                 /*use_single_precision=*/false,
+                                                                 /*vector_width=*/2);
+            std::regex exp_decl(R"(declare <2 x double> @llvm\.exp\.v2f64\(<2 x double>\))");
+            std::regex exp_call(R"(call <2 x double> @llvm\.exp\.v2f64\(<2 x double> .*\))");
+            REQUIRE(std::regex_search(no_library_module_str, m, exp_decl));
+            REQUIRE(std::regex_search(no_library_module_str, m, exp_call));
+
+            // Check exponential calls are replaced with calls to SVML library.
+            std::string svml_library_module_str = run_llvm_visitor(nmodl_text,
+                                                                   /*opt=*/false,
+                                                                   /*use_single_precision=*/false,
+                                                                   /*vector_width=*/2,
+                                                                   /*vec_lib=*/"SVML");
+            std::regex svml_exp_decl(R"(declare <2 x double> @__svml_exp2\(<2 x double>\))");
+            std::regex svml_exp_call(R"(call <2 x double> @__svml_exp2\(<2 x double> .*\))");
+            REQUIRE(std::regex_search(svml_library_module_str, m, svml_exp_decl));
+            REQUIRE(std::regex_search(svml_library_module_str, m, svml_exp_call));
+            REQUIRE(!std::regex_search(svml_library_module_str, m, exp_call));
+
+            // Check that supported exponential calls are replaced with calls to MASSV library (i.e.
+            // operating on vector of width 2).
+            std::string massv2_library_module_str = run_llvm_visitor(nmodl_text,
+                                                                     /*opt=*/false,
+                                                                     /*use_single_precision=*/false,
+                                                                     /*vector_width=*/2,
+                                                                     /*vec_lib=*/"MASSV");
+            std::regex massv2_exp_decl(R"(declare <2 x double> @__expd2_P8\(<2 x double>\))");
+            std::regex massv2_exp_call(R"(call <2 x double> @__expd2_P8\(<2 x double> .*\))");
+            REQUIRE(std::regex_search(massv2_library_module_str, m, massv2_exp_decl));
+            REQUIRE(std::regex_search(massv2_library_module_str, m, massv2_exp_call));
+            REQUIRE(!std::regex_search(massv2_library_module_str, m, exp_call));
+
+            // Check no replacement for MASSV happens for non-supported vector widths.
+            std::string massv4_library_module_str = run_llvm_visitor(nmodl_text,
+                                                                     /*opt=*/false,
+                                                                     /*use_single_precision=*/false,
+                                                                     /*vector_width=*/4,
+                                                                     /*vec_lib=*/"MASSV");
+            std::regex exp4_call(R"(call <4 x double> @llvm\.exp\.v4f64\(<4 x double> .*\))");
+            REQUIRE(std::regex_search(massv4_library_module_str, m, exp4_call));
+
+            // Check correct replacement of @llvm.exp.v4f32 into @vexpf when using Accelerate.
+            std::string accelerate_library_module_str =
+                run_llvm_visitor(nmodl_text,
+                                 /*opt=*/false,
+                                 /*use_single_precision=*/true,
+                                 /*vector_width=*/4,
+                                 /*vec_lib=*/"Accelerate");
+            std::regex accelerate_exp_decl(R"(declare <4 x float> @vexpf\(<4 x float>\))");
+            std::regex accelerate_exp_call(R"(call <4 x float> @vexpf\(<4 x float> .*\))");
+            std::regex fexp_call(R"(call <4 x float> @llvm\.exp\.v4f32\(<4 x float> .*\))");
+            REQUIRE(std::regex_search(accelerate_library_module_str, m, accelerate_exp_decl));
+            REQUIRE(std::regex_search(accelerate_library_module_str, m, accelerate_exp_call));
+            REQUIRE(!std::regex_search(accelerate_library_module_str, m, fexp_call));
         }
     }
 }

--- a/test/unit/codegen/codegen_llvm_ir.cpp
+++ b/test/unit/codegen/codegen_llvm_ir.cpp
@@ -1068,22 +1068,19 @@ SCENARIO("Vector library calls", "[visitor][llvm][vector_lib]") {
             NEURON {
                 SUFFIX hh
                 NONSPECIFIC_CURRENT il
-                RANGE minf, mtau
             }
             STATE {
                 m
             }
             ASSIGNED {
                 v (mV)
-                minf
-                mtau (ms)
             }
             BREAKPOINT {
                 SOLVE states METHOD cnexp
                 il = 2
             }
             DERIVATIVE states {
-                m' = (minf-m) / mtau
+                m = exp(m)
             }
         )";
 

